File: mcp/stoat_mcp/guards.py
"""The deterministic blocks.

Pure functions, no I/O beyond path resolution, so they are exhaustively
testable. Every one is enforced regardless of what the MCP client does,
because the protocol guarantees nothing useful here: human-in-the-loop is a
SHOULD, tool annotations are explicitly advisory, and elicitation is a
capability a client may simply not declare. What stoat wants guaranteed,
stoat enforces.

See docs/design/mcp-server.md §4 and json-contract-draft.md §7.
"""

from __future__ import annotations

import os
import re
import time
from pathlib import Path

from .errors import GuardRejection

# A VM name is a DIRECTORY name under the data root. That is the whole reason
# this is strict: the name becomes a path component, and stoat resolves every
# operation by directory. Letters, digits, dash, underscore, dot, but never a
# leading dot (which would hide it and collide with stoat's own bookkeeping)
# and never "." or ".." (which are traversal).
_VM_NAME = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")

# Catalog image IDs look like "alpine-virt" or "ubuntu-24.04". Anything with a
# separator is a path, and a path is what section 7.1 #4 forbids: an absolute
# BYO path is an arbitrary host file read, booted as a disk.
_IMAGE_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")

_RECIPE_NAME = re.compile(r"^[a-z][a-z0-9-]*$")
_RECIPE_REF = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._/-]*$")


def check_vm_name(name: str) -> str:
    """Reject anything that is not a plain VM name.

    A name reaches stoat as a directory component, so "../../etc" or an
    absolute path would escape the data root. Rejecting by pattern rather
    than by sanitizing is deliberate: there is no correct way to "fix" a
    malicious name, and a sanitizer that silently rewrites input hides the
    attempt.
    """
    if not isinstance(name, str) or not name.strip():
        raise GuardRejection("vm name is required")
    if name != name.strip():
        raise GuardRejection(f"vm name {name!r} has leading or trailing whitespace")
    if name in (".", ".."):
        raise GuardRejection(f"vm name {name!r} is a path traversal")
    if os.sep in name or (os.altsep and os.altsep in name):
        raise GuardRejection(f"vm name {name!r} contains a path separator")
    if "\x00" in name:
        raise GuardRejection("vm name contains a null byte")
    if not _VM_NAME.match(name):
        raise GuardRejection(
            f"vm name {name!r} is not allowed: use letters, digits, dot, dash "
            "and underscore, starting with a letter or digit"
        )
    return name


def check_image_id(image: str) -> str:
    """Catalog IDs only. A path is refused outright.

    `create --image /abs/path.qcow2` is an arbitrary host file read booted as
    a disk (json-contract-draft.md section 7.1 #4). The CLI accepts it because
    a human bringing their own image is a supported workflow; an agent
    choosing one is not.
    """
    if not isinstance(image, str) or not image.strip():
        raise GuardRejection("image id is required")
    if os.sep in image or (os.altsep and os.altsep in image) or image.startswith("~"):
        raise GuardRejection(
            f"image {image!r} looks like a path. Only catalog image ids are "
            "accepted; run list_images to see them."
        )
    if not _IMAGE_ID.match(image):
        raise GuardRejection(f"image id {image!r} is not a valid catalog id")
    return image


def check_index_name(ref: str) -> str:
    """Accept an index recipe name with an optional safe git ref.

    The recipe name is always resolved through the curated index. A ref may
    contain slashes for branches such as ``feature/topic`` but cannot contain
    URL, option, or path-traversal syntax.
    """
    if not isinstance(ref, str) or not ref or ref != ref.strip():
        raise GuardRejection("recipe index name is required")
    if ref.startswith("-") or "\\" in ref or "\x00" in ref:
        raise GuardRejection(f"recipe index name {ref!r} is not safe")
    if ref.count("@") > 1:
        raise GuardRejection(f"recipe index name {ref!r} has more than one ref")
    name, separator, branch = ref.partition("@")
    if not _RECIPE_NAME.fullmatch(name):
        raise GuardRejection(
            f"recipe index name {ref!r} must start with a letter and contain only letters, digits and dashes"
        )
    if not separator:
        return ref
    if not _RECIPE_REF.fullmatch(branch) or branch.startswith("/") or branch.endswith("/"):
        raise GuardRejection(f"recipe ref {branch!r} is not safe")
    parts = branch.split("/")
    if ".." in branch or any(
        part in ("", ".", "..")
        or part.startswith(".")
        or part.endswith(".")
        or part.endswith(".lock")
        for part in parts
    ):
        raise GuardRejection(f"recipe ref {branch!r} is not safe")
    return ref


def shared_dir(vm: str, data_root: str | os.PathLike[str] | None = None) -> Path:
    """The one host directory an agent may read or write for this VM.

    Mirrors the layout stoat itself uses for the writable 9p export
    (~/.stoat/shared/<vm>/), which is what makes a copy in or out equivalent
    to the guest writing into its own work share.
    """
    root = Path(data_root) if data_root else Path(os.environ.get("STOAT_HOME") or Path.home() / ".stoat")
    return (root / "shared" / check_vm_name(vm)).resolve()


def check_host_path(path: str, vm: str, data_root: str | os.PathLike[str] | None = None) -> str:
    """Confine a host path to ~/.stoat/shared/<vm>/, resolving symlinks first.

    The order matters and is the whole guard: resolve, THEN compare. A check
    performed before resolution is defeated by a symlink inside the sandbox
    pointing out of it, and such a symlink is trivially created by the guest,
    which has that same directory mounted read-write.

    Comparison is by path PARTS, not by string prefix. "~/.stoat/shared/work"
    is a string prefix of "~/.stoat/shared/work-evil", so a prefix test would
    admit a sibling directory belonging to a different VM.

    This check is possible here, unlike for the 9p share itself, precisely
    because stoat performs the copy: a 9p operation is served by QEMU and
    stoat never sees it.
    """
    if not isinstance(path, str) or not path.strip():
        raise GuardRejection("path is required")
    if "\x00" in path:
        raise GuardRejection("path contains a null byte")

    sandbox = shared_dir(vm, data_root)
    # expanduser first: "~" is the user's, not ours, and os.path.realpath
    # would treat a literal "~" as a relative directory name.
    candidate = Path(os.path.expanduser(path))
    if not candidate.is_absolute():
        raise GuardRejection(
            f"path {path!r} must be absolute, under {sandbox}"
        )
    resolved = Path(os.path.realpath(candidate))

    if resolved != sandbox and sandbox not in resolved.parents:
        raise GuardRejection(
            f"path {path!r} resolves to {resolved}, which is outside this VM's "
            f"shared directory ({sandbox}). Only files under that directory can "
            "be copied in or out."
        )
    return str(resolved)


class RateLimiter:
    """A token bucket per tool name, plus one bucket every tool shares.

    The MCP spec makes rate limiting a server MUST. The numbers here are a
    starting point rather than a tuned value: generous enough that ordinary
    agent work never notices, tight enough that a runaway loop calling
    `create` a thousand times stops being the host's problem.

    The shared bucket is what bounds the server as a whole. Per-tool buckets
    alone let a caller burst `capacity` times against each of ~20 tools, so
    the real ceiling was 20x the number anyone read off this class.

    Not thread-safe on purpose; the stdio server is single-threaded, and a
    lock here would imply a concurrency story that does not exist.
    """

    def __init__(
        self,
        capacity: int = 30,
        refill_per_second: float = 0.5,
        total_capacity: int = 60,
        total_refill_per_second: float = 2.0,
    ) -> None:
        if capacity < 1 or total_capacity < 1:
            raise ValueError("capacity must be at least 1, or nothing can ever run")
        if refill_per_second <= 0 or total_refill_per_second <= 0:
            # A bucket that never refills is a bucket that permanently bricks
            # the tool after `capacity` calls. Refusing at construction beats
            # discovering it in production, and it is why the message below
            # can divide by this without a guard.
            raise ValueError("refill_per_second must be positive")
        self.capacity = capacity
        self.refill_per_second = refill_per_second
        self.total_capacity = total_capacity
        self.total_refill_per_second = total_refill_per_second
        self._buckets: dict[str, tuple[float, float]] = {}

    def check(self, tool: str, now: float | None = None) -> None:
        """Consume one token for `tool` and one shared token.

        Both buckets are read before either is charged. A call refused by one
        must not spend from the other: a hot tool hitting its own limit would
        otherwise drain the shared bucket and starve every other tool.
        """
        now = time.monotonic() if now is None else now
        tool_tokens = self._read(tool, self.capacity, self.refill_per_second, now, tool)
        total_tokens = self._read(
            "", self.total_capacity, self.total_refill_per_second, now, "the server"
        )
        self._buckets[tool] = (tool_tokens - 1.0, now)
        self._buckets[""] = (total_tokens - 1.0, now)

    def _read(
        self, key: str, capacity: int, refill: float, now: float, subject: str
    ) -> float:
        """Return key's token count at `now`, raising when it is below one."""
        tokens, last = self._buckets.get(key, (float(capacity), now))
        tokens = min(capacity, tokens + (now - last) * refill)
        if tokens < 1.0:
            wait = (1.0 - tokens) / refill
            raise GuardRejection(
                f"rate limit reached for {subject}; retry in about {wait:.0f}s"
            )
        return tokens


def check_flag_free(values: list[str], what: str) -> list[str]:
    """Refuse a value that kong would read as a flag.

    `forward` and `check_recipes` splat their list arguments into argv as
    positionals. `forward(pairs=["--clear"])` reached kong as the --clear
    flag and wiped the VM's forwards, from a call that passed clear=False.
    Nothing escapes the process (there is no shell), so this is argv
    confusion, and refusing a leading dash closes it.
    """
    for v in values:
        if not isinstance(v, str) or not v.strip():
            raise GuardRejection(f"{what} contains an empty value")
        if v.startswith("-"):
            raise GuardRejection(f"{what} value {v!r} may not start with a dash")
    return values


def strip_forbidden(patch: dict[str, object]) -> dict[str, object]:
    """Remove parameters an agent may never set, from a patch it supplied.

    `share` is the one that matters: it grants an arbitrary host directory,
    read-write, into a guest, and core.Patch is exactly the generic-map shape
    that makes it reachable from JSON tool arguments
    (json-contract-draft.md section 7.1 #2).

    Dropped silently rather than rejected, because an agent copying a VM
    object it read back into an update call is doing something reasonable;
    the field simply has no effect. Reading `share` is fine, setting it is not,
    and that asymmetry is deliberate.
    """
    return {k: v for k, v in patch.items() if k not in FORBIDDEN_PATCH_KEYS}


# Never accepted as tool input, at any level. These are absent rather than
# gated: a parameter that exists will eventually be reachable.
FORBIDDEN_PATCH_KEYS = frozenset({"share", "image", "base", "iso", "console_password"})

File: internal/mcpsrv/guards.go
// Package mcpsrv serves the MCP protocol from the stoat binary. Every tool
// runs the guards in this file, calls internal/core, and returns a wire DTO.
package mcpsrv

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strings"

	"github.com/novusedge/stoat/internal/config"
	"github.com/novusedge/stoat/internal/qemu"
)

// A VM name becomes a directory name under the data root, so the pattern is
// what keeps an operation inside it. Rejecting beats sanitizing: a rewrite
// hides the attempt.
var vmNameRE = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._-]*$`)

// A catalog image id never contains a separator. An absolute path here is an
// arbitrary host file read, booted as a disk.
var imageIDRE = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._-]*$`)

var (
	indexNameRE = regexp.MustCompile(`^[a-z0-9][a-z0-9-]*$`)
	gitRefRE    = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._/-]*$`)
	paramNameRE = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)
	svcNameRE   = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._@-]*$`)
)

// forbiddenPatchKeys are never accepted as tool input at any level. share
// grants an arbitrary host directory into a guest read-write.
var forbiddenPatchKeys = []string{"share", "image", "base", "iso", "console_password"}

func checkVMName(name string) (string, error) {
	if strings.TrimSpace(name) == "" {
		return "", fmt.Errorf("vm name is required")
	}
	if name != strings.TrimSpace(name) {
		return "", fmt.Errorf("vm name %q has leading or trailing whitespace", name)
	}
	if name == "." || name == ".." {
		return "", fmt.Errorf("vm name %q is a path traversal", name)
	}
	if strings.ContainsAny(name, `/\`) {
		return "", fmt.Errorf("vm name %q contains a path separator", name)
	}
	if strings.ContainsRune(name, 0) {
		return "", fmt.Errorf("vm name contains a null byte")
	}
	if !vmNameRE.MatchString(name) {
		return "", fmt.Errorf("invalid vm name %q: must match %s", name, vmNameRE)
	}
	return name, nil
}

func checkImageID(image string) (string, error) {
	if strings.TrimSpace(image) == "" {
		return "", fmt.Errorf("image id is required")
	}
	if strings.ContainsAny(image, `/\`) || strings.HasPrefix(image, "~") {
		return "", fmt.Errorf("image %q looks like a path; only catalog image ids are accepted, run list_images to see them", image)
	}
	if !imageIDRE.MatchString(image) {
		return "", fmt.Errorf("image id %q is not a valid catalog id", image)
	}
	return image, nil
}

// sharedDir is the one host directory an agent may read or write for a VM.
// It mirrors the layout of the writable 9p export.
func sharedDir(vm string) (string, error) {
	name, err := checkVMName(vm)
	if err != nil {
		return "", err
	}
	return filepath.EvalSymlinks(filepath.Join(config.Root(), "shared", name))
}

// checkHostPath confines a host path to ~/.stoat/shared/<vm>/. The order is
// the guard: resolve symlinks, then compare. A check before resolution is
// defeated by a symlink the guest creates inside its own mounted share.
// Comparison is by path element, so the sibling "work-evil" cannot pass as
// "work".
func checkHostPath(path, vm string) (string, error) {
	if strings.TrimSpace(path) == "" {
		return "", fmt.Errorf("path is required")
	}
	if strings.ContainsRune(path, 0) {
		return "", fmt.Errorf("path contains a null byte")
	}
	sandbox, err := sharedDir(vm)
	if err != nil {
		return "", err
	}
	// Expand ~ first. filepath.EvalSymlinks reads a literal "~" as a
	// relative directory name.
	candidate := config.Expand(path)
	if !filepath.IsAbs(candidate) {
		return "", fmt.Errorf("path %q must be absolute, under %s", path, sandbox)
	}
	resolved, err := resolveExisting(candidate)
	if err != nil {
		return "", err
	}
	if resolved != sandbox && !strings.HasPrefix(resolved, sandbox+string(os.PathSeparator)) {
		return "", fmt.Errorf("path %q resolves to %s, which is outside this VM's shared directory (%s)", path, resolved, sandbox)
	}
	return resolved, nil
}

// resolveExisting resolves every symlink in p. A destination that does not
// exist yet is legitimate for copy_from, so the deepest existing ancestor is
// resolved and the remaining elements are rejoined.
func resolveExisting(p string) (string, error) {
	rest := ""
	for cur := filepath.Clean(p); ; {
		if r, err := filepath.EvalSymlinks(cur); err == nil {
			return filepath.Join(r, rest), nil
		}
		parent := filepath.Dir(cur)
		if parent == cur {
			return "", fmt.Errorf("path %q cannot be resolved", p)
		}
		rest = filepath.Join(filepath.Base(cur), rest)
		cur = parent
	}
}

// checkFlagFree refuses a value kong would read as a flag. forward and
// check_recipes splat their list arguments into argv as positionals, and
// forward(pairs=["--clear"]) once reached kong as --clear and wiped a VM's
// forwards.
func checkFlagFree(values []string, what string) error {
	for _, v := range values {
		if strings.TrimSpace(v) == "" {
			return fmt.Errorf("%s contains an empty value", what)
		}
		if strings.HasPrefix(v, "-") {
			return fmt.Errorf("%s value %q may not start with a dash", what, v)
		}
	}
	return nil
}

// stripForbidden drops keys an agent may never set. It returns a new map:
// an agent that reads a VM back and passes it to update is doing something
// reasonable, and the field simply has no effect.
func stripForbidden(patch map[string]any) map[string]any {
	out := make(map[string]any, len(patch))
	for k, v := range patch {
		if !slices.Contains(forbiddenPatchKeys, k) {
			out[k] = v
		}
	}
	return out
}

// checkIndexName splits "<name>" or "<name>@<ref>" from the recipe index. A
// URL is refused here rather than in core: add_recipe is the tool an agent
// reaches, and a URL is a repository nobody curated.
func checkIndexName(ref string) (string, string, error) {
	if strings.TrimSpace(ref) == "" {
		return "", "", fmt.Errorf("recipe name is required")
	}
	if strings.ContainsAny(ref, ":/\\") {
		return "", "", fmt.Errorf("invalid recipe name %q: index names only, not a URL", ref)
	}
	name, gitRef, hasRef := strings.Cut(ref, "@")
	if strings.Contains(gitRef, "@") {
		return "", "", fmt.Errorf("invalid recipe name %q: at most one @ref", ref)
	}
	if !indexNameRE.MatchString(name) {
		return "", "", fmt.Errorf("invalid recipe name %q: must match %s", name, indexNameRE)
	}
	if hasRef {
		if !gitRefRE.MatchString(gitRef) || strings.Contains(gitRef, "..") {
			return "", "", fmt.Errorf("invalid ref %q for recipe %q", gitRef, name)
		}
	}
	return name, gitRef, nil
}

func checkParamName(name string) (string, error) {
	if !paramNameRE.MatchString(name) {
		return "", fmt.Errorf("invalid param name %q: must match %s", name, paramNameRE)
	}
	return name, nil
}

// checkGuestPath requires an absolute guest path. A relative path is an
// error and is never resolved against $HOME, so one tool call means one
// thing on every guest.
func checkGuestPath(path string) (string, error) {
	if strings.TrimSpace(path) == "" {
		return "", fmt.Errorf("guest path is required")
	}
	if strings.ContainsRune(path, 0) {
		return "", fmt.Errorf("guest path contains a null byte")
	}
	if !strings.HasPrefix(path, "/") {
		return "", fmt.Errorf("guest path %q must be absolute", path)
	}
	return path, nil
}

// checkSvcName bounds a service name. svc and svc_status render the guest
// file's template and pass the name as $1, so this is depth rather than the
// boundary.
func checkSvcName(name string) (string, error) {
	if !svcNameRE.MatchString(name) {
		return "", fmt.Errorf("invalid service name %q: must match %s", name, svcNameRE)
	}
	return name, nil
}

// requireRunning gates a tool whose description promises it refuses on a
// stopped VM. Without it, sshx.Run against a stopped VM's forwarded port
// surfaces ssh's own connection-refused exit rather than this error.
func requireRunning(v *config.VM) error {
	if !qemu.Running(v) {
		return fmt.Errorf("%w: %s", qemu.ErrNotRunning, v.Name)
	}
	return nil
}
